tcmur_cmd_handler.c

```diff
@@ -33,33 +34,42 @@ static void _cleanup_spin_lock(void *arg)
 }
 
 void tcmur_command_complete(struct tcmu_device *dev, struct tcmulib_cmd *cmd,
-			    int rc)
+			    int rc, bool timeout)
 {
 	struct tcmur_device *rdev = tcmu_get_daemon_dev_private(dev);
 
 	pthread_cleanup_push(_cleanup_spin_lock, (void *)&rdev->lock);
-	pthread_spin_lock(&rdev->lock);
-
-	tcmulib_command_complete(dev, cmd, rc);
+	if (!timeout && cmd->timer)
+		tcmu_del_timer(cmd->timer);
 
+	pthread_spin_lock(&rdev->lock);
+	tcmulib_command_complete(dev, cmd, rc, timeout);
 	pthread_spin_unlock(&rdev->lock);
 	pthread_cleanup_pop(0);
 }
 
 static void aio_command_finish(struct tcmu_device *dev, struct tcmulib_cmd *cmd,
-			       int rc)
+			       int rc, bool timeout)
 {
 	struct tcmur_device *rdev = tcmu_get_daemon_dev_private(dev);
 	int wake_up;
 
-	tcmur_command_complete(dev, cmd, rc);
+	tcmur_command_complete(dev, cmd, rc, timeout);
 	track_aio_request_finish(rdev, &wake_up);
 	while (wake_up) {
 		tcmulib_processing_complete(dev);
 		track_aio_wakeup_finish(rdev, &wake_up);
 	}
 }
 
+void tcmur_cmd_timeout(struct tcmu_timer *timer, void *data)
+{
+	struct tcmulib_cmd *cmd = (struct tcmulib_cmd *)data;
+	struct tcmu_device *dev = cmd->dev;
+
```

Comment:
For glfs, do you need to perform any cleanup in the handler or is ok to leave it running?

Reply:
No I meant this function will clean up the tcmu-runner/libtcmu bits but it leaves the command running in glfs. Do you want to call into glfs.c and kill the command in there?

For example, if you ran each command in a thread and cmd1 is in thread1, it's timer fires and libtcmu/runner fail the command, but it is still running on thread1. The initiator gets the failed command decides to retry. That cmd instance is run on thread2 at the same time thread1 is still running. Thread2 completes, then the initiator sends a new command to the same sectors so now the old command on thread1 is running at the same time this new command is. Thread1 might then overwrite newer data.

I was just asking if glfs handles those type of things already or if you need to call into glfs.c to tell it to stop/kill the command that is running on thread1?

Reply:
For a similar problem but in tcmu-runner core, I think we need to add checks to the compound commands's done functions.

For example, for a xcopy we could timeout during the read part of the command, return failure to the initiator, but then glfs returns success from the read part, calls the done function and then starts the write part while possibly another xcopy is now started.

For CAW we could have another problem where the read part timeout, we return failure to the initiator, but we leave the caw_lock held so future CAWs are still blocked on that stuck original CAW.

Reply:
> No I meant this function will clean up the tcmu-runner/libtcmu bits but it leaves the command running in glfs. Do you want to call into glfs.c and kill the command in there?
> 
> For example, if you ran each command in a thread and cmd1 is in thread1, it's timer fires and libtcmu/runner fail the command, but it is still running on thread1. The initiator gets the failed command decides to retry. That cmd instance is run on thread2 at the same time thread1 is still running. Thread2 completes, then the initiator sends a new command to the same sectors so now the old command on thread1 is running at the same time this new command is. Thread1 might then overwrite newer data.
> 
> I was just asking if glfs handles those type of things already or if you need to call into glfs.c to tell it to stop/kill the command that is running on thread1?

Yeah, while there is one GLUSTER ping timeout in glusterfs backend, I am still investigating whether will the IO threads be cleaned after this timeout. We meant to set the timeout value here will be larger than the GLUSTER ping timeout value.

Reply:
> For a similar problem but in tcmu-runner core, I think we need to add checks to the compound commands's done functions.
> 
> For example, for a xcopy we could timeout during the read part of the command, return failure to the initiator, but then glfs returns success from the read part, calls the done function and then starts the write part while possibly another xcopy is now started.
> 
> For CAW we could have another problem where the read part timeout, we return failure to the initiator, but we leave the caw_lock held so future CAWs are still blocked on that stuck original CAW.

This also make sense. Will fix it later.